### ISSUE TITLE
Crash(?) requiring system restart

### ISSUE THREAD

**Issue by lukef** [flutter/flutter#11124]
Not sure how to classify this issue or how to reproduce it but i'll try to provide as much info as possible. 

System versions:

macOS: 10.12.5
IntelliJ: Ultimate 2017.2 EAP
Chrome" 59.0.3071.115 (Official Build) (64-bit)
Flutter: master branch

Sometimes my system will get into a state where it requires a restart to keep working. Whenever I try to run the `flutter` tool I will get the following message:

`flutter: fork: Resource temporarily unavailable`

When this message appears, the device chooser will disappear from intellij and Chrome also stops responding and starts showing the "sad face" tab whenever a new tab is opened.

See below for screenshot of missing device chooser.

<img width="590" alt="screen shot 2017-07-08 at 1 26 09 pm" src="https://user-images.githubusercontent.com/37941/27988927-c5ad15f2-63e2-11e7-8a7a-fe52f72f20db.png">

**Comment by branflake2267** [comment]
This has happened to me twice.

**Comment by lukef** [comment]
Adding some extra info:

Talked to @collinjackson and he said it is definitely a simulator-related issue. The following workaround is from Collin:

Run `ps x | grep DTServiceHub` to confirm you have a bunch of instances and then run `killall DTServiceHub`. 

Also, a ton of logs will appear similar to `DTServiceHub[1137]: Could not create service named com.apple.instruments.server.services.processcontrol.posixspawn` before the whole thing gives up.

**Comment by collinjackson** [comment]
I haven't see the issue occur for me after running `chmod 0000 /Applications/Xcode.app/Contents/SharedFrameworks/DVTInstrumentsFoundation.framework/Resources/DTServiceHub`

Could just be a coincidence, or IntelliJ + Flutter is leaking DTServiceHub processes somehow.

**Comment by devoncarew** [comment]
Hmm, looks like I have 488 background processes running DTServiceHub.

**Comment by lukef** [comment]
With my current configuration, if I open Xcode (8.3.3) and just start running an app (non-flutter, objc) in the simulator then I see the same issue. It could be an Apple issue.

**Comment by devoncarew** [comment]
Good to know. As a workaround, when launching an app on the simulator, we could check for lots of DTServiceHub processes running (> 100?), and killall them. It's not too pretty, but users wouldn't see the crashing issue anymore in the flutter daemon, chrome, ...

**Comment by branflake2267** [comment]
I see this exception occur in the IDE as well. 

## Version information

IntelliJ IDEA `2017.1.5` • Flutter plugin `15.1` • Dart plugin `171.4694.29`

## Exception

Flutter device daemon #3: process exited during startup. Exit code: 128, stderr:
/Users/branflake2267/git/flutter/bin/flutter: fork: Resource temporarily unavailable


```
com.intellij.execution.ExecutionException: Flutter device daemon #3: process exited during startup. Exit code: 128, stderr:
/Users/branflake2267/git/flutter/bin/flutter: fork: Resource temporarily unavailable

	at io.flutter.run.daemon.DeviceDaemon$Command.start(DeviceDaemon.java:222)
	at io.flutter.run.daemon.DeviceService.chooseNextDaemon(DeviceService.java:220)
	at io.flutter.utils.Refreshable.runInBackground(Refreshable.java:214)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by collinjackson** [comment]
It's important to note that once your system has gotten into a bad state (lots of DTServiceHub processes failing), all sorts of things will start failing. I couldn't even run `ls` on the command line sometimes. And of course Xcode gets unhappy as well and can't run Flutter apps or anything else.

@lukef, are you sure you were able to repro DTServiceHub leak without using Flutter at all -- you rebooted your system and didn't do any Flutter/IntelliJ development and ran out of processes? It seems more likely to me that your system was already in a bad state when you tried to run the non Flutter app in Xcode.

**Comment by lukef** [comment]
@collinjackson - you're correct. I thought i had restarted before testing but i just tried it all again and I am not able to replicate the DTServiceHub issue after restarting on a non-flutter project. It must be related to flutter.

**Comment by eseidelGoogle** [comment]
Presumably this is related to our migration to using instruments instead of `ios-deploy`?  @cbracken?

**Comment by cbracken** [comment]
> Presumably this is related to our migration to using instruments instead of ios-deploy?

Looks like it, though I'm still trying to identify the exact conditions under which the processes leak. Repeatedly running `flutter devices` triggers short-lived `DTServiceHub` processes that drop away within seconds.

However, once I start a simulator, I see that running `flutter devices` triggers:
1. the usual short-lived process (`/Applications/Xcode.app/Contents/SharedFrameworks/DVTInstrumentsFoundation.framework/Resources/DTServiceHub`) followed by
1. an instance that starts but never terminates from the simulator (`/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/Developer/Library/PrivateFrameworks/DVTInstrumentsFoundation.framework/DTServiceHub`)

Running `flutter devices` again with the simulator still up appears to sometimes trigger a hung process. Closing the simulator seems to clear all of these out.

TL;DR this does appear to be the result of some sort of interaction between the instruments-based device lookup and the simulator, and I've only been able to reproduce this leak when a simulator is up, but I still haven't nailed down the root cause. Will update when I've got more.

**Comment by cbracken** [comment]
Mitigation is now in place that should significantly reduce the chance of `instruments`-related issues when a simulator is up, as well as some precautionary cleanup of orphaned processes should they occur. Please re-open if you continue to see this.
